src/modules/bulletproofs/main_impl.h

```diff
@@ -0,0 +1,237 @@
+     * `secp256k1_bulletproof_generators_create`; stored in this structure to allow consistent
+     * generators between functions using `secp256k1_bulletproof_generators` and functions
+     * using the Pedersen commitment module. */
+    secp256k1_ge *blinding_gen;
+};
+
+#include "modules/bulletproofs/inner_product_impl.h"
+#include "modules/bulletproofs/rangeproof_impl.h"
+#include "modules/bulletproofs/util.h"
+
+secp256k1_bulletproof_generators *secp256k1_bulletproof_generators_create(const secp256k1_context *ctx, const secp256k1_generator *blinding_gen, size_t n) {
+    secp256k1_bulletproof_generators *ret;
+    secp256k1_rfc6979_hmac_sha256 rng;
+    unsigned char seed[64];
+    secp256k1_gej precompj;
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(blinding_gen != NULL);
+
+    ret = (secp256k1_bulletproof_generators *)checked_malloc(&ctx->error_callback, sizeof(*ret));
+    if (ret == NULL) {
+        return NULL;
+    }
+    ret->gens = (secp256k1_ge *)checked_malloc(&ctx->error_callback, (n + 1) * sizeof(*ret->gens));
+    if (ret->gens == NULL) {
+        free(ret);
+        return NULL;
+    }
+    ret->blinding_gen = &ret->gens[n];
+    ret->n = n;
+
+    secp256k1_fe_get_b32(&seed[0], &secp256k1_ge_const_g.x);
+    secp256k1_fe_get_b32(&seed[32], &secp256k1_ge_const_g.y);
+
+    secp256k1_rfc6979_hmac_sha256_initialize(&rng, seed, 64);
+    for (i = 0; i < n; i++) {
+        unsigned char tmp[32] = { 0 };
+        secp256k1_generator gen;
+        secp256k1_rfc6979_hmac_sha256_generate(&rng, tmp, 32);
+        CHECK(secp256k1_generator_generate(ctx, &gen, tmp));
+        secp256k1_generator_load(&ret->gens[i], &gen);
+
+        secp256k1_gej_set_ge(&precompj, &ret->gens[i]);
+    }
+
+    secp256k1_generator_load(&ret->blinding_gen[0], blinding_gen);
+    secp256k1_gej_set_ge(&precompj, &ret->blinding_gen[0]);
+
+    return ret;
+}
+
+void secp256k1_bulletproof_generators_destroy(const secp256k1_context* ctx, secp256k1_bulletproof_generators *gens) {
+    (void) ctx;
+    if (gens != NULL) {
+        free(gens->gens);
+        free(gens);
+    }
+}
+
+int secp256k1_bulletproof_rangeproof_verify(const secp256k1_context* ctx, secp256k1_scratch_space *scratch, const secp256k1_bulletproof_generators *gens, const unsigned char *proof, size_t plen,
+ const uint64_t *min_value, const secp256k1_pedersen_commitment* commit, size_t n_commits, size_t nbits, const secp256k1_generator *value_gen, const unsigned char *extra_commit, size_t extra_commit_len) {
+    int ret;
+    size_t i;
+    secp256k1_ge *commitp;
+    secp256k1_ge value_genp;
+    const secp256k1_ge *commitp_ptr;
+    const uint64_t *minvalue_ptr;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(scratch != NULL);
+    ARG_CHECK(gens != NULL);
+    ARG_CHECK(gens->n >= 2 * nbits * n_commits);
+    ARG_CHECK(proof != NULL);
+    ARG_CHECK(commit != NULL);
+    ARG_CHECK(n_commits > 0);
```

Comment:
We should additionally have a upper bound for `n_commits` because it's used to calculate the size of many allocs and the number of points in the multi-exponentiation. And maybe a comment that explains that the bound is there for a good reason.